### ISSUE TITLE
shutil.copyfileobj stopped working in 1.2.0

### ISSUE THREAD

**Issue by henryiii** [python/mypy#15031]
**Bug Report**

MyPy 1.2.0 doesn't like `shutil.copyfileobj` anymore, with "can't infer type argument 1".

Hit in https://github.com/pypa/cibuildwheel/pull/1468.

**To Reproduce**

```python
import shutil
import typing

stdin = typing.cast(typing.IO[bytes], None)
with open("random_file.txt", "rb") as from_file:
    shutil.copyfileobj(from_file, stdin)
```

**Expected Behavior**

```console
$ pipx run mypy==1.1.1 --strict tmp.py
Success: no issues found in 1 source file
```

https://mypy-play.net/?mypy=1.1.1&python=3.11&gist=7f9fbe8e5df5f52e5c639e30671eb208

**Actual Behavior**

```console
$ pipx run mypy==1.2.0 --strict tmp.py
tmp.py:6: error: Cannot infer type argument 1 of "copyfileobj"  [misc]
Found 1 error in 1 file (checked 1 source file)
```

https://mypy-play.net/?mypy=latest&python=3.11&gist=dbc914004646a0c0ad90926c501d1daa

**Your Environment**

- Mypy version used: 1.1.1 & 1.2.0
- Mypy command-line flags: (Seen in playground with nothing)
- Mypy configuration options from `mypy.ini` (and other config files): (Seen in playground with nothing)
- Python version used: 3.11 and others.

**Comment by hauntsaninja** [comment]
Seems caused by change in typeshed: https://github.com/python/mypy/pull/14902/files#diff-363460977156fcfda748f21565484fe1d5862edf2823e9784cbf34d1e52ff2f2

**Comment by boennecd** [comment]
Also fails with Python 3.10.

**Comment by henryiii** [comment]
Should I open an issue over there, then?

**Comment by JelleZijlstra** [comment]
I think this is actually a mypy bug, though one that was exposed through the typeshed change. It would help to try to create a reproducer that doesn't rely on the stdlib (basically, copy-paste in the definitions of `IO`, `open`, and `copyfileobj`, and simplify them as much as possible).

**Comment by AlexWaygood** [comment]
It looks like #14975 fixes this false positive, which strongly implies to me that this is the same bug as #14943

**Comment by AlexWaygood** [comment]
Here's a minimal-ish repro:

```py
from typing import AnyStr, Generic, Protocol, TypeVar, overload

T_co = TypeVar("T_co", covariant=True)
T_contra = TypeVar("T_contra", contravariant=True)

class SupportsRead(Protocol[T_co]):
    def read(self) -> T_co: ...

class SupportsWrite(Protocol[T_contra]):
    def write(self, s: T_contra) -> object: ...

def copyfileobj(fsrc: SupportsRead[AnyStr], fdst: SupportsWrite[AnyStr]) -> None: ...
    
class WriteToMe(Generic[AnyStr]):
    @overload
    def write(self: WriteToMe[str], s: str) -> int: ...
    @overload
    def write(self: WriteToMe[bytes], s: bytes) -> int: ...

class WriteToMeOrReadFromMe(WriteToMe[AnyStr]):
    def read(self) -> AnyStr: ...

copyfileobj(WriteToMeOrReadFromMe[bytes](), WriteToMe[bytes]())
```

It does indeed end up looking very similar to #14943.

**Comment by henryiii** [comment]
Thanks! Type ignore removed in https://github.com/pypa/cibuildwheel/pull/2054.
